src/algorithm/const_fold.rs

```diff
@@ -0,0 +1,302 @@
+        })
+        .flatten()
+}
+
+/// Attempt to evaluate and generate rewrites for the operation at `op_node`
+fn fold_op(
+    hugr: &impl HugrView,
+    op_node: Node,
+    reg: &ExtensionRegistry,
+) -> Option<(SimpleReplacement, Vec<RemoveConstIgnore>)> {
+    let (in_consts, removals): (Vec<_>, Vec<_>) = hugr
+        .node_inputs(op_node)
+        .filter_map(|in_p| {
+            let (con_op, load_n) = get_const(hugr, op_node, in_p)?;
+            Some(((in_p, con_op), RemoveConstIgnore(load_n)))
+        })
+        .unzip();
+    let neighbour_op = hugr.get_optype(op_node);
+    // attempt to evaluate op
+    let folded = fold_const(neighbour_op, &in_consts)?;
+    let (op_outs, consts): (Vec<_>, Vec<_>) = folded.into_iter().unzip();
+    let nu_out = op_outs
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, out)| {
+            // map from the ports the op was linked to, to the output ports of
+            // the replacement.
+            hugr.single_linked_input(op_node, out)
+                .map(|np| (np, i.into()))
+        })
+        .collect();
+    let replacement = const_graph(consts, reg);
+    let sibling_graph = SiblingSubgraph::try_from_nodes([op_node], hugr)
+        .expect("Operation should form valid subgraph.");
+
+    let simple_replace = SimpleReplacement::new(
+        sibling_graph,
+        replacement,
+        // no inputs to replacement
+        HashMap::new(),
+        nu_out,
+    );
+    Some((simple_replace, removals))
+}
+
+/// If `op_node` is connected to a LoadConstant at `in_p`, return the constant
+/// and the LoadConstant node
+fn get_const(hugr: &impl HugrView, op_node: Node, in_p: IncomingPort) -> Option<(Const, Node)> {
+    let (load_n, _) = hugr.single_linked_output(op_node, in_p)?;
+    let load_op = hugr.get_optype(load_n).as_load_constant()?;
+    let const_node = hugr
+        .linked_outputs(load_n, load_op.constant_port())
+        .exactly_one()
+        .ok()?
+        .0;
+
+    let const_op = hugr.get_optype(const_node).as_const()?;
+
+    // TODO avoid const clone here
+    Some((const_op.clone(), load_n))
+}
+
+/// Exhaustively apply constant folding to a HUGR.
+pub fn constant_fold_pass(h: &mut impl HugrMut, reg: &ExtensionRegistry) {
+    loop {
+        // would be preferable if the candidates were updated to be just the
+        // neighbouring nodes of those added.
+        let rewrites = find_consts(h, h.nodes(), reg).collect_vec();
+        if rewrites.is_empty() {
+            break;
+        }
+        for (replace, removes) in rewrites {
+            h.apply_rewrite(replace).unwrap();
+            for rem in removes {
+                if let Ok(const_node) = h.apply_rewrite(rem) {
+                    // if the LoadConst was removed, try removing the Const too.
+                    if h.apply_rewrite(RemoveConst(const_node)).is_err() {
+                        // const cannot be removed - no problem
+                        continue;
```

Comment:
I find it surprising that this is not removed by the rewrite itself. Maybe add a comment about this in the docstring of `find_consts`?